File: Search/binary_search.py
import csv
import random
import time

def load_dataset(filename):
    data = []
    with open(filename, mode='r') as file:
        reader = csv.reader(file)
        for row in reader:
            data.append(int(row[0]))
    return data

def binary_search(data, target):
    start_time = time.perf_counter()
    left = 0
    right = len(data) - 1

    while left <= right:
        mid = (left + right) // 2
        number = data[mid]

        if number == target:
            return (time.perf_counter() - start_time) * 1000
        if number < target:
            left = mid + 1
        else:
            right = mid - 1

    return (time.perf_counter() - start_time) * 1000

n_list = [1000, 5000, 10000, 50000, 100000, 500000, 1000000, 5000000, 10000000, 30000000]

for n in n_list:
    print(f"Processing dataset with {n} records")
    dataset_filename = f"dataset/quick_sort_result/quick_sort_{n}.csv"
    output_filename = f"dataset/binary_search_result/binary_search_{n}.txt"

    data = load_dataset(dataset_filename)

    target_best = data[(len(data) - 1) // 2]
    target_avg = random.choice(data)
    target_worst = -1

    best_case_time = binary_search(data, target_best)
    avg_case_time = binary_search(data, target_avg)
    worst_case_time = binary_search(data, target_worst)

    with open(output_filename, 'w') as file:
        file.write(f"Best case time   : {best_case_time:.5f} ms\n")
        file.write(f"Average case time: {avg_case_time:.5f} ms\n")
        file.write(f"Worst case time  : {worst_case_time:.5f} ms\n")

File: Search/BinarySearch.java
import java.io.*;
import java.util.*;

public class BinarySearch {

    public static List<long[]> loadDataset(String filename) {
        List<long[]> data = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                long number = Long.parseLong(parts[0]);
                data.add(new long[]{number});
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return data;
    }

    public static double binarySearch(List<long[]> data, long target) {
        long startTime = System.nanoTime();
        int left = 0;
        int right = data.size() - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            long number = data.get(mid)[0];

            if (number == target) {
                return (System.nanoTime() - startTime) / 1000000.0;
            }
            if (number < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return (System.nanoTime() - startTime) / 1000000.0;
    }

    public static void main(String[] args) {
        int[] nList = {1000, 5000, 10000, 50000, 100000, 500000, 1000000, 5000000, 10000000, 30000000};

        for (int n : nList) {
            System.out.println("Processing dataset with " + n + " records");

            String datasetFilename = "dataset/quick_sort_result/quick_sort_" + n + ".csv";
            String outputFilename = "dataset/binary_search_result/binary_search_" + n + ".txt";

            List<long[]> data = loadDataset(datasetFilename);

            Random rand = new Random();

            long target_best = data.get((data.size() - 1) / 2)[0];
            long target_avg = data.get(rand.nextInt(data.size()))[0];
            long target_worst = -1;

            double bestCaseTime = binarySearch(data, target_best);
            double avgCaseTime = binarySearch(data, target_avg);
            double worstCaseTime = binarySearch(data, target_worst);

            System.out.printf("Best case time   : %.5f ms%n", bestCaseTime);
            System.out.printf("Average case time: %.5f ms%n", avgCaseTime);
            System.out.printf("Worst case time  : %.5f ms%n%n", worstCaseTime);

            try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFilename))) {
                writer.write(String.format("Best case time   : %.5f ms\n", bestCaseTime));
                writer.write(String.format("Average case time: %.5f ms\n", avgCaseTime));
                writer.write(String.format("Worst case time  : %.5f ms\n", worstCaseTime));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        
    }
}
